### PR TITLE
Add OverheadIcons, preliminary dialog.

### PR DESCRIPTION
**Pull request by b1gd3vd0g**
Overhead Icons can be shown over entities to convey messages to the player easily (for example, i have new dialog!)
Preliminary Dialog Bubbles are included as well. Also a couple bug fixes, a little clean-up (sorting/alphabetizing), and added Chad's face to the tab in your browser (tested in firefox)!

### CODE REVIEW

File: js/classes/entities/foreground/dialog_bubble.js [hunk 1]
```diff
@@ -1,147 +1,207 @@
 /**
- * The section of dialog that is displayed near the speaker.
- * @author Nathan Hinthorne
+ * A DialogBubble is a unit of an entity's conversation array. It contains a portion of 
+ * text, which is being said by a specified speaker (or not, as in the case of gameplay instruction).
+ * It <will be> displayed at the bottom of the screen, and show the Speaker's head animation above it.
+ * 
+ * THIS CLASS IS NOT COMPLETE!
+ * 
+ * @author Nathan Hinthorne, Devin Peevy
```
Comment (tclaar) [inline js/classes/entities/foreground/dialog_bubble.js:8]: I think multiple authors should be put on separate lines each with their own @author tag, e.g.
```
@author Nathan Hinthorne
@author Devin Peevy
```

File: index.html [hunk 2]
```diff
@@ -2,6 +2,10 @@
 <html>
 
 <head>
+    <!-- THIS LINE IS GOING TO SOLVE AN ERROR FROM FIREFOX.
+    TODO: It is a good idea for somebody to put this in though! The icon on the tab of the browser! -->
```
Comment (calebkrauter) [inline index.html:6]: Nice! I use FireFox!

File: js/classes/entities/foreground/dialog_bubble.js [hunk 3]
```diff
@@ -1,147 +1,207 @@
...
-        CTX.font = FONT.VT323_NORMAL;
-        CTX.fillStyle = "black";
-
-        // read in the text and draw it line by line
-        let lines = this.text.split("\n");
-        for (let i = 0; i < lines.length; i++) {
-            CTX.fillText(lines[i], this.pos.x + 10, this.pos.y - 110 + (i * 20));
-        }
-    };
+// Nathan wrote a lot of good stuff down here, and I expect to steal quite a bit of his code
```
Comment (NathanHinthorne) [inline js/classes/entities/foreground/dialog_bubble.js:57]: Fine by me!

File: js/classes/entities/foreground/dialog_bubble.js [hunk 4]
```diff
@@ -1,147 +1,207 @@
...
 
-    /**
-     * Removes this DialogBubble from the world.
+    /** 
+     * A JSON object with all of the characters who talk. It stores Animators of their heads.
+     * The speaker field of any DialogBubble should be a constant from this getter.
      */
-    remove() {
-        this.removedFromWorld = true;
+    static get SPEAKERS() {
```
Comment (NathanHinthorne) [inline js/classes/entities/foreground/dialog_bubble.js:41]: This getter adds a lot more modularity to the class. I like it!

File: js/classes/entities/foreground/dialog_bubble.js [hunk 5]
```diff
@@ -1,147 +1,207 @@
...
-     * @param {string} type The type of dialog bubble. DialogBubble.NORMAL, .THOUGHT, or .SHOUT.
+     * Constructs a Dialog Bubble 
+     * @param {Animator} speaker The person saying the dialog. Should be a constant from DialogBubble.SPEAKERS.If nobody, use null.
+     * @param {string} text 
+     * @param {Array[Choice]} choices 
+     * @param {boolean} endOfConversation 
      */
-    constructor(speaker, text, type) {
-        /** The speaker object (should be an npc or enemy). */
+    constructor(speaker, text, endOfConversation = false, choices = []) {
```
Comment (calebkrauter) [inline js/classes/entities/foreground/dialog_bubble.js:18]: Good idea on endOfConversation!

File: js/classes/entities/foreground/dialog_bubble.js [hunk 6]
```diff
@@ -1,147 +1,207 @@
...
-            case DialogBubble.SHOUT:
-                return 160;
+        this.isEnd = endOfConversation;
+        this.choices = choices;
+        // Note to devs: Choice objects are coming soon. Kinda my last priority as far as dialog comes.
+        switch (this.speaker) {
+            case DialogBubble.SPEAKERS.CHAD:
+                this.type = DialogBubble.LEFT_BUBBLE;
+                break;
+            case null:
```
Comment (calebkrauter) [inline js/classes/entities/foreground/dialog_bubble.js:29]: This looks great for now, wondering if we will have more cases or not.

File: js/classes/entities/foreground/dialog_bubble.js (hunk 4 above)
Comment (calebkrauter) [inline js/classes/entities/foreground/dialog_bubble.js:41]: Is this for displaying speakers? Are we going to list all of the possible speakers here? Hoping that we can make it more generic if it gets too big. @B1GD3VD0G

File: js/classes/entities/foreground/overhead_icon.js
Comment (NathanHinthorne) [inline js/classes/entities/foreground/overhead_icon.js:1]: Looks great! It's simple and does the job well.

File: js/classes/entities/foreground/dialog_bubble.js [hunk 7]
```diff
@@ -1,147 +1,207 @@
...
+//         } else {
+//             // large box
+//             return 5;
+//         }
+//     };
+
+//     /**
+//      * Updates the x and y position of the dialog bubble.
+//      */
+//     update() {
```
Comment (calebkrauter) [inline js/classes/entities/foreground/dialog_bubble.js:143]: Don't we need an update function in this class? This is commented out, will it break the game? @B1GD3VD0G

File: js/classes/entities/foreground/dialog_bubble.js [hunk 8]
```diff
@@ -1,147 +1,207 @@
...
+//         // this.x = this.speaker.x;
+//         // this.y = this.speaker.y-10;
+//     };
+
+//     draw() {
```
Comment (calebkrauter) [inline js/classes/entities/foreground/dialog_bubble.js:148]: Don't we need a draw function @B1GD3VD0G ?

File: js/classes/entities/midground/npc/village/papa_chad.js
Comment (NathanHinthorne) [inline js/classes/entities/midground/npc/village/papa_chad.js:1]: Thanks for taking out the user input to papa chad. It'll help for his transition to being an npc

File: js/scripts/dimension/playground_dimension.js
Comment (NathanHinthorne) [inline js/scripts/dimension/playground_dimension.js:1]: Good change. I like it!

File: js/classes/entities/foreground/overhead_icon.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
+/**
+ * An OverheadIcon is a symbol which floats above an entity, indicating some sort of message to the player.
+ * 
+ * DETERMINED MEANINGS:
+ * 
+ * TRIANGLE:    dialog. [GREEN = quest, WHITE = new, unseen].
```
Comment (calebkrauter) [inline js/classes/entities/foreground/overhead_icon.js:6]: I like this, good idea!

File: js/classes/entities/foreground/overhead_icon.js [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        );
+        /** The position of this OverheadIcon in the game world. */
+        this.pos = Vector.add(this.target.pos, aboveAndCenter);
+
+        this.animator = new Animator(OverheadIcon.SPRITESHEET,
+            new Vector(this.color * OverheadIcon.SIZE.x, this.type * OverheadIcon.SIZE.y),
+            OverheadIcon.SIZE,
+            1, 1);
+    };
+
```
Comment (tclaar) [inline js/classes/entities/foreground/overhead_icon.js:52]: I know its pretty obvious what they're for, but the functions and getters below should probably have JSDoc.

**Review by NathanHinthorne — APPROVED** [review]
Good additions @B1GD3VD0G !
The new triangle for each new piece of dialog is a great idea

File: js/classes/entities/foreground/overhead_icon.js [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+class OverheadIcon {
+    
+    /**
+     * A constructor for an overhead icon.
+     * @param {Entity} target This is the Entity above which the Icon will be drawn.
+     * @param {number} targetWidth This is the width of the target, so that it can be placed correctly.
+     * @param {number} type OverheadIcon.TRIANGLE, .QUESTION or .EXCLAMATION - determines the shape of the icon.
+     * @param {number} color OverheadIcon.WHITE, .GREEN, .YELLOW, or .RED - determines the color of the icon.
+     */
+    constructor(target, targetWidth, type, color) {
```
Comment (calebkrauter) [inline js/classes/entities/foreground/overhead_icon.js:21]: Just a suggestion, could we be more specific with the type like iconType or something like that?

File: js/classes/entities/foreground/overhead_icon.js [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+        /** The entity above which this OverheadIcon should be drawn. */
+        this.target = target;
+        /** The width of the target entity, in pixels. */
+        this.targetWidth = targetWidth; // It's unfortunate that we need this, but since size is static we do.
+        /** The number representing the type constant of this OverheadIcon. */
+        this.type = type;
+        /** The number representing the color constant of this OverheadIcon. */
+        this.color = color;
+
+        const aboveAndCenter = new Vector(
```
Comment (calebkrauter) [inline js/classes/entities/foreground/overhead_icon.js:40]: It appears that this is the positioning code for putting an icon above an npc or CHAD, is that correct? Does it update properly if the npc is moving or do we even want that? Are we planning on npcs and CHAD being stationary during dialog?

File: js/classes/entities/foreground/overhead_icon.js [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+
+        this.animator = new Animator(OverheadIcon.SPRITESHEET,
+            new Vector(this.color * OverheadIcon.SIZE.x, this.type * OverheadIcon.SIZE.y),
+            OverheadIcon.SIZE,
+            1, 1);
+    };
+
+    update() {
+        // Add this vector to target's pos to get its pos.
+        const aboveAndCenter = new Vector(
```
Comment (calebkrauter) [inline js/classes/entities/foreground/overhead_icon.js:55]: Why are we creating this constant before and inside of update? @B1GD3VD0G

File: js/classes/entities/midground/npc/village/papa_chad.js [hunk 14]
```diff
@@ -56,190 +46,29 @@ class PapaChad {
...
-                } else {
-                    this.canDash = true;
-                    this.hasDashed = false;
-                    this.isDashing = true;
-
-
-                }
+        
+        // Set the velocity, according to gravity.
+        this.velocity = {
```
Comment (calebkrauter) [inline js/classes/entities/midground/npc/village/papa_chad.js:58]: Nice!

**Review by calebkrauter — APPROVED** [review]
Looks good! Just some questions.

File: js/scripts/dimension/playground_dimension.js [hunk 15]
```diff
@@ -91,7 +91,7 @@ const loadPlaygroundDimension = () => {
         // NOTE: we can't activate music until the user has interacted with the canvas. (this issue is inherent to HTML5)
```
Comment (tclaar) [inline js/scripts/dimension/playground_dimension.js:91]: Lines 91-101 look like they are duplicates of lines 79-89.

**Review by tclaar — APPROVED** [review]
Looks good! Just a few minor comments.

File: js/classes/entities/foreground/dialog_bubble.js (hunk 7 above)
Comment (b1gd3vd0g) [reply js/classes/entities/foreground/dialog_bubble.js:143]: havent gotten there yet. class is incomplete.

File: js/classes/entities/foreground/dialog_bubble.js (hunk 8 above)
Comment (b1gd3vd0g) [reply js/classes/entities/foreground/dialog_bubble.js:148]: yes we do.

File: js/classes/entities/foreground/dialog_bubble.js (hunk 1 above)
Comment (b1gd3vd0g) [reply js/classes/entities/foreground/dialog_bubble.js:8]: Really? That's dope and makes sense, thank you!

File: js/classes/entities/foreground/overhead_icon.js (hunk 10 above)
Comment (b1gd3vd0g) [reply js/classes/entities/foreground/overhead_icon.js:52]: Will do

File: js/scripts/dimension/playground_dimension.js (hunk 15 above)
Comment (b1gd3vd0g) [reply js/scripts/dimension/playground_dimension.js:91]: You're right. This wasn't my class, this was something that I was fixing to get rid of a red error bar in the firefox debugger.

**Pull request merged**